### ISSUE TITLE
define `Base.length` and `Base.endof`

### ISSUE THREAD

**Issue by ExpandingMan** [quinnj/WeakRefStrings.jl#25]
Currently, on 0.4.1, if you call `length(str)` on a `WeakRefString` you get
```
ERROR: you must implement endof(WeakRefString{UInt8})
```
One of the problems this causes is that DataFrames containing `WeakRefString`s don't display correctly.

**Comment by nalimilan** [comment]
When do you get DataFrames containing `WeakRefString`s? `WeakRefStringArray` columns are supposed to return `String` entries when indexing.

**Comment by ExpandingMan** [comment]
Oh I see why then.  I'm working on an [implementation](github.com/ExpandingMan/Arrow.jl) of the Arrow format, and I have some array objects which basically just contain pointers and construct objects on demand.  This can be done either with `WeakRefString`s or `String`s.  When these objects are set to return `String`s they are essentially very similar to `WeakRefStringArray`.  There wouldn't necessarily be any reason to expose these arrow arrays to the user returning `WeakRefString`s.

Still, `WeakRefString` should have these methods anyway, no?

**Comment by nalimilan** [comment]
I'm not sure. I suspect @quinnj has avoided implementing most string methods precisely to prevent people from using this unsafe type directly.

**Comment by ExpandingMan** [comment]
Yeah I figured that.  Still, i'm not sure that leaving methods undefined that are supposed to be defined is the best way of doing that.  After all, people still wind up using `WeakRefString` somehow.

**Comment by nalimilan** [comment]
> After all, people still wind up using WeakRefString somehow.

Precisely, they shouldn't. README says:
> Usage of WeakRefStrings is discouraged for general users. Currently, a WeakRefString purposely does not implement many Base Julia String interface methods due to many recent changes to Julia's builtin String interface, as well as the complexity to do so correctly.

**Comment by ExpandingMan** [comment]
All I meant was that `WeakRefString`s may occur in contexts where one is not doing "string manipulation" per se, but basic methods such as `length` get called, e.g. the above.  But point taken, I have no strong feelings about it one way or another, let's see what @quinnj thinks and then close the issue.

**Comment by iamed2** [comment]
We (@spurll) ran into this problem just using CSV.jl.

**Comment by nalimilan** [comment]
@iamed2 How exactly? AFAIK you're not supposed to ever see a `WeakRefString` object when working with CSV.jl (only `WeakRefStringArray`).

**Comment by spurll** [comment]
If you’d like a minimal example I can provide one on Monday.

**Comment by nalimilan** [comment]
Then this should be filed against CSV.jl.

**Comment by quinnj** [comment]
Yes, I'd certainly like to hear what happened, as @nalimilan said, you should basically never see WeakRefStrings themselves from CSV anymore (apart from a WeakRefString*Array* which, when indexed, gives Strings)
